File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<vector>
#include <queue>
#include <cstring>
#define Y first
#define X second
using namespace std;
int dy[4] = {0,1,0,-1};
int dx[4] = {1,0,-1,0};
int solution(vector<vector<int> > maps)
{
    int answer = -1;
    int vis[101][101];
    memset(vis,-1,sizeof(vis));
    queue<pair<int,int>> Q;
    Q.push({0,0});
    vis[0][0] = 1;
    while (!Q.empty()){
        auto nq = Q.front();
        Q.pop();
        for(int i=0;i<4;i++){
            int ny = nq.Y + dy[i];
            int nx = nq.X + dx[i];
            if (ny < 0 || ny >= maps.size() || nx < 0 || nx >= maps[0].size()) continue;
            if (vis[ny][nx] != -1) continue;
            if (maps[ny][nx] == 0) continue;
            vis[ny][nx] = vis[nq.Y][nq.X]+1;
            Q.push({ny,nx});
        }
    }
    return vis[maps.size()-1][maps[0].size()-1];
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;
class Solution {
    static final int[] dy = {1,0,-1,0};
    static final int[] dx = {0,-1,0,1};
    public int solution(int[][] maps) {
        int n = maps.length;
        int m = maps[0].length;
        Queue<int[]> Q = new LinkedList<>();
        boolean vis[][] = new boolean[n][m];
        Q.add(new int[]{0,0,1});
        vis[0][0] = true;
        while(!Q.isEmpty()){
            int[] q = Q.poll();
            int x=q[0];
            int y=q[1];
            int cnt = q[2];
            for(int i=0;i<4;i++){
                int nx = dx[i] + x;
                int ny = dy[i] + y;
                if (nx < 0 || ny < 0 || nx >= m || ny >= n) continue;
                if (vis[ny][nx] == true || maps[ny][nx] == 0) continue;
                vis[ny][nx] = true;
                if (ny == n-1 && nx == m-1)
                    return cnt+1;
                Q.add(new int[]{nx,ny,cnt+1});
            }
        }
        return -1;
    }
}